### ISSUE TITLE
Rating Assignments

### ISSUE THREAD

**Issue by shindavid** [shindavid/AlphaZeroArcade#59]
Each run of the AlphaZero process produces a series of progressively stronger agents. When we wish to experiment with some change in the process, whether a code-change, model-architecture change, or hyperparameter change, we launch a new run [1]. It is important to have some rigorous way to compare the two runs. How can we do this? More generally, if we launch R runs, how do we identify the "best" run?

For Connect4, we had a perfect oracle that could objectively identify each move as correct or incorrect. We could use this to plot a mistake-rate graph for each run, and compare runs by comparing graphs.

In games where we don't have a perfect oracle, we need something else. We need to rely on win-loss records that result when we pit these agents against each other. But what does that look like exactly? If we have R runs, each of which has produced G generations of agents, how many of the (R*G choose 2) possible matches do we need to run? The fewer, the better, from a computational budget perspective - how few can we get away with? Which ones should we run, to get the most bang for our computational buck?

One idea is to consider each run in isolation. Pit the G agents against each other, creating enough win/loss data to assign ratings to them. Then compare runs by comparing rating-curves. Ratings can be computed via the [Bradley-Terry rating model](https://en.wikipedia.org/wiki/Bradley%E2%80%93Terry_model) [2].

BUT - can ratings computed from different populations be compared to each other meaningfully? There is some reason to believe that maybe they can. We can fix a rating of 0 to the Random agent, which just picks a move uniformly at random among all possible moves. If we normalize to the Elo scale, a rating of delta of 100 corresponds to a ~76% probability of winning. So if run 1 yields a rating 400 player, it's a player that has a 1 - (1-0.76)^4 probability of beating a random player. If run 2 yields a rating 400 player, it also has that same probability of beating the same random player. It may be reasonable to say that the two players are equivalent in skill.

On the other hand, maybe if you pit these two rating-400 players against each other, one player dominates the other. This can happen because **ratings are relative to a population of players** - an agent might have a rating of r1 in population P1, and a rating of r2 in population P2, and theoretically we cannot make any guarantees about the relationship between r1 and r2.

So maybe we have to run some matches between agents from different runs. Again, how many do we need?

We are in need of a **rating-assignment-process**. It should run alongside an AlphaZero run, identifying when new agents are produced, scheduling matches between the agents, and computing ratings from win/loss records. It should have access to past runs if necessary. Ideally, it has these properties:

* Can scale up/down based on how many machines/cores we dedicate to it (particularly if we run on cloud).
* Can be interrupted/restarted arbitrarily while preserving state
* Intelligently chooses next match to run based on anticipated information-gain from result
* Knows that within a run, the agents are _expected_ (but not guaranteed) to be progressively stronger, which allows for some sort of interpolation/regularization
* Answers useful questions about the nature of ratings across populations - do ratings empirically appear population-independent, for the populations produced by the AlphaZero process?
* Comes with useful web-browser tools to interactively visualize outputs

The script `py/assign_ratings.py` currently serves as a first-whack at this task.

[1] Alternatively, we can create a checkpoint for a run, and then spawn new runs from that checkpoint to experiment with various changes. This is cheaper than doing a new run for scratch for every experiment. David Wu, the KataGo author, told me that this strategy was very important in his work.

[2] In the Bradley-Terry model, each agent's skill level is estimated by a real number, and the (logit) probability that one agent defeats another is modeled as the difference of their skill levels. Given a set of match records, we can compute Bradley-Terry ratings via MLE very quickly and accurately. Note that more well-known rating systems like Elo and TrueSkill are basically just variants of Bradley-Terry that weight more recent matches more heavily, to deal with the reality that human agents evolve in their skill level over time. When dealing with static computer agents, we don't need to worry about skill evolution, and so there is no reason to go beyond simple Bradley-Terry.

**Comment by shindavid** [comment]
I'm happy with the solution I arrived at, which is simply to have a game specific reference agent family parameterized by a single numerical value to score against. Simple and avoids all the headaches discussed above.
